crates/core/src/idx/planner/mod.rs

```diff
@@ -60,13 +69,53 @@ impl<'a> StatementContext<'a> {
 			order: stm.order(),
 			cond: stm.cond(),
 			group: stm.group(),
+			is_perm,
 		})
 	}
 
+	pub(crate) async fn check_table_permission(
+		&self,
+		tb: &str,
+	) -> Result<GrantedPermission, Error> {
+		if !self.is_perm {
+			return Ok(GrantedPermission::Full);
+		}
+		// Get the table for this planner
+		match self.ctx.tx().get_tb(self.ns, self.db, tb).await {
+			Ok(table) => {
+				// TODO(tobiemh): we should really
+				// not even get here if the table
+				// permissions are NONE, because
+				// there is no point in processing
+				// a table which we can't access.
+				let perms = self.stm.permissions(&table, false);
+				// If permissions are specific, we
+				// need to fetch the record content.
+				if perms.is_specific() {
+					return Ok(GrantedPermission::Specific);
+				}
+				// If permissions are NONE, we also
+				// need to fetch the record content.
+				if perms.is_none() {
+					return Ok(GrantedPermission::None);
+				}
+			}
+			Err(Error::TbNotFound {
+				..
+			}) => {
+				// We can safely ignore this error,
+				// as it just means that there is no
+				// table and no permissions defined.
```

Comment:
Well then we would not need to iterate at all! You cannot have records without a table defined :D